hbase/src/main/java/magnolify/hbase/HBaseType.scala

```diff
@@ -0,0 +1,267 @@
+        .asScala
+        .map { case (qualifier, value) => new String(qualifier, UTF_8) -> value }
+        .toMap
+    )
+
+  def apply(v: T, row: Array[Byte], family: Array[Byte], ts: Long = 0L): Mutation = to(v)
+    .foldLeft(new Put(row)) { case (row, (qualifier, value)) =>
+      row.addColumn(family, qualifier.getBytes(UTF_8), ts, value)
+    }
+}
+
+object HBaseType {
+  implicit def apply[T: HBaseField]: HBaseType[T] = HBaseType(CaseMapper.identity)
+
+  def apply[T](cm: CaseMapper)(implicit f: HBaseField[T]): HBaseType[T] = f match {
+    case r: HBaseField.Record[_] =>
+      new HBaseType[T] {
+        private val caseMapper: CaseMapper = cm
+
+        override def from(xs: Map[String, Array[Byte]]): T = r.get(xs, null)(caseMapper).get
+
+        override def to(v: T): Map[String, Array[Byte]] = r.put(null, v)(caseMapper)
+      }
+    case _ =>
+      throw new IllegalArgumentException(s"BigtableType can only be created from Record. Got $f")
+  }
+}
+
+sealed trait HBaseField[T] extends Serializable {
+  def get(xs: Map[String, Array[Byte]], k: String)(cm: CaseMapper): Value[T]
+  def put(k: String, v: T)(cm: CaseMapper): Map[String, Array[Byte]]
+}
+
+object HBaseField {
+  sealed trait Record[T] extends HBaseField[T]
+
+  sealed trait Primitive[T] extends HBaseField[T] {
+    def size: Option[Int]
+    def fromBytes(v: Array[Byte]): T
+    def toBytes(v: T): Array[Byte]
+
+    override def get(xs: Map[String, Array[Byte]], k: String)(cm: CaseMapper): Value[T] =
+      xs.get(k) match {
+        case Some(v) => Value.Some(fromBytes(v))
+        case None    => Value.None
+      }
+
+    override def put(k: String, v: T)(cm: CaseMapper): Map[String, Array[Byte]] =
+      Map(k -> toBytes(v))
+  }
+
+  // ////////////////////////////////////////////////
+  type Typeclass[T] = HBaseField[T]
+
+  def join[T](caseClass: CaseClass[Typeclass, T]): HBaseField[T] = {
+    if (caseClass.isValueClass) {
+      val p = caseClass.parameters.head
+      val tc = p.typeclass
+      new HBaseField[T] {
+        override def get(xs: Map[String, Array[Byte]], k: String)(cm: CaseMapper): Value[T] =
+          tc.get(xs, k)(cm).map(x => caseClass.construct(_ => x))
+
+        override def put(k: String, v: T)(cm: CaseMapper): Map[String, Array[Byte]] =
+          p.typeclass.put(k, p.dereference(v))(cm)
+      }
+    } else {
+      new Record[T] {
+        private def qualifier(prefix: String, label: String): String =
+          if (prefix == null) label else s"$prefix.$label"
+
+        override def get(xs: Map[String, Array[Byte]], k: String)(cm: CaseMapper): Value[T] = {
+          var fallback = true
+          val r = caseClass.construct { p =>
+            val q = qualifier(k, cm.map(p.label))
+            val v = p.typeclass.get(xs, q)(cm)
+            if (v.isSome) {
+              fallback = false
+            }
+            v.getOrElse(p.default)
+          }
+          // result is default if all fields are default
```

Comment:
Unless I'm misunderstanding the `construct` lambda, it looks like `result` defaults if even one field is default?

Reply:
I copied this from the `bigtable` module. I don't get either why we use the `Value` wrapper. I'd see If I can clean that

Reply:
See https://github.com/spotify/magnolify/pull/847